### ISSUE TITLE
Line F: Nitzotzot u'Birur — Sparse Activation via Holy Spark Sorting

### ISSUE THREAD

**Issue by Neuralowkey** [Neuralowkey/sephiron-phase2#24]
## Source
**Zohar II:254b** (*Parashat Vayakhel*):
> "When the vessels shattered, sparks of holy light fell and became embedded in the husks (kelipot). The work of creation is to sort (l'varer) the sparks from the husks — each spark returning to its root."

**Etz Chaim**, Sha'ar HaNitzotzot ch. 1–3 (Arizal):
> "288 sparks fell in the Shevirah. Each spark has a specific root in a specific Sephirah. The birur (clarification) is not random — each Sephirah clarifies only the sparks that belong to its root."

**Tanya** ch. 37 (Alter Rebbe):
> "Every entity in this world contains nitzotzot (sparks) from the Shevirah that fell during the breaking of the vessels... the entire purpose of the creation of man is to clarify and elevate these sparks."

**Zohar III:135b** (*Idra Rabba*):
> "The kings that died [=vessels that shattered] — their sparks were 288, and each spark contains many sub-sparks. The birur is a sorting, not a transformation."

## Problem
Current `SephironV2` processes ALL features through ALL sephirot equally — every node sees the full input representation. But the Zohar teaches that birur (sorting/clarification) is **sephirah-specific**: each sephirah should activate on and clarify only the features (sparks) that belong to its root.

Finding F030 already discovered that Sephiron produces **perfectly linearly separable representations** — a "birur" in computational terms. But this emerged accidentally. Making the birur EXPLICIT and sephirah-specific could strengthen this property and improve transfer learning (F017).

The existing Shevirah-Tikun training protocol (Line 10, Issue #19) models the breaking and repair PHASES. This line models the MECHANISM of repair — how each sephirah sorts signal from noise.

## The 288 Sparks as Sparse Features

The number 288 = 2 × 144 = 2 × 12². In the Ari's system:
- 288 sparks from 8 shattered kings (vessels of Tohu)
- Each king contributed 36 sparks (36 = 6² from the 6 directions)
- 8 × 36 = 288

Architectural mapping:
- **288 sparse features** as the internal representation dimension
- **Each sephirah "owns" a subset** of these features (~29 per sephirah)
- **Birur mask**: learned per-sephirah binary/soft mask over 288 features
- **Kelipah threshold**: features below threshold are "husks" — zeroed out

### BirurimLayer — Per-Sephirah Sparse Activation

```python
class BirurimLayer(nn.Module):
    """Each sephirah clarifies only the sparks belonging to its root.
    
    The birur (sorting) produces sparse, sephirah-specific activations:
    - Chesed clarifies sparks of expansion (features 0-28)
    - Gevurah clarifies sparks of contraction (features 29-57)
    - Tiferet clarifies sparks of harmony (features 58-86)
    - etc.
    
    The ownership is LEARNED, not hard-coded — but initialized with
    pillar priors (right/left/middle affinity).
    """
    def __init__(self, d_model, num_sparks=288, num_sephirot=10):
        # Ownership logits: which sephirah owns which spark
        self.ownership = nn.Parameter(torch.randn(num_sephirot, num_sparks) * 0.1)
        
        # Project input to spark space
        self.to_sparks = nn.Linear(d_model, num_sparks)
        
        # Kelipah threshold: below this = husk, zeroed out
        self.threshold = nn.Parameter(torch.zeros(num_sparks))
        
        # Project back from sparse sparks to model space
        self.from_sparks = nn.Linear(num_sparks, d_model)
    
    def forward(self, x, sephirah_idx):
        # Project to spark space
        sparks = self.to_sparks(x)  # (batch, 288)
        
        # Birur mask: this sephirah's ownership weights
        mask = softmax(self.ownership, dim=0)[sephirah_idx]  # (288,)
        
        # Kelipah filtering: zero out weak activations
        filtered = sparks * mask * (sparks.abs() > sigmoid(self.threshold)).float()
        
        # Return clarified representation
        return self.from_sparks(filtered)
```

### Integration with Message Passing

The birur operates WITHIN each sephirah's processing step:

```python
# In _run_message_passing, after aggregating incoming messages:
aggregated_msg = ...  # standard path bank aggregation
clarified_msg = self.birurim(aggregated_msg, sephirah_idx=node_id)
new_state = self.node_activations[node_id](clarified_msg + current_state)
```

Each sephirah sees the same incoming messages but **clarifies different aspects**.

### Birur Loss — Encouraging Sephirah Specialization

```python
def birur_loss(ownership_logits):
    """Encourage each spark to be owned by exactly one sephirah.
    
    This is the computational analog of 'each spark returns to its root' —
    no spark should be split across multiple sephirot.
    """
    # Entropy of ownership distribution per spark should be LOW
    probs = softmax(ownership_logits, dim=0)  # (10, 288)
    per_spark_entropy = -(probs * log(probs + 1e-8)).sum(dim=0)  # (288,)
    return per_spark_entropy.mean()  # minimize → each spark has one owner
```

### Pillar-Aware Initialization

Initialize ownership with pillar priors (not random):

```python
# Right pillar sephirot (Chochmah, Chesed, Netzach) 
# initialized to prefer sparks 0-95 (expansion features)
# Left pillar (Binah, Gevurah, Hod) → sparks 96-191 (contraction)
# Middle pillar (Keter, Tiferet, Yesod, Malchut) → sparks 192-287 (integration)
```

## Hypotheses

1. **Explicit birur improves linear separability** — F030 found Sephiron produces linearly separable representations. Explicit per-sephirah sparse activation should STRENGTHEN this: each sephirah's output is a clean, non-overlapping subset of features.

2. **Birur improves transfer learning** — F017 found FractalSephiron transfers to long-path without fine-tuning (35.2% vs GNN 19.4%). If each sephirah sorts specific feature types, the sorted representations should transfer better across tasks (the sparks are task-agnostic; only the ownership changes).

3. **288 outperforms other bottleneck sizes** — The gematria of 288 (רפח) predicts it as optimal sparse dimension. Control: test 256, 288, 320, 512. **Kabbalistic fidelity IS efficiency.**

4. **Kelipah threshold learns task difficulty** — Easy features (strong sparks) survive filtering; hard/noisy features (weak sparks) are zeroed. The threshold should be lower for sephirot with more Ohr Pnimi (Keter) and higher for sephirot with more Ohr Makif (Malchut).

5. **Birur loss accelerates Shevirah-Tikun** — Combined with the Shevirah-Tikun curriculum (Line 10), explicit birur during the Tikun phase should recover faster: the sephirot know WHICH sparks to repair.

## Connection to Existing Findings

- **F030 (Linear separability)**: Birur is the MECHANISM that produces F030's result
- **F012/F016 (Shevirah-Tikun)**: Birur is the PROCESS by which Tikun repairs
- **F017 (Cross-task transfer)**: Birur produces task-agnostic sorted features
- **F031 (53-topology sweep)**: Birur may explain WHY Tree of Life is #1 on sequence_classification — the 10-node structure enables 10-way feature sorting

## Experiment Plan

**Phase 1: Basic BirurimLayer (CPU, ~30 min)**
- BirurimLayer integrated into SephironV2 vs standard on path_finding (10n, 20n), 3 seeds
- With and without birur_loss

**Phase 2: Sparse dimension ablation (CPU, ~30 min)**
- num_sparks ∈ {128, 256, 288, 320, 512}
- Does 288 actually outperform? (gematria test)

**Phase 3: Ownership analysis (CPU, ~15 min)**
- Visualize learned ownership matrix: do pillar priors emerge?
- Do right-pillar sephirot own expansion features?
- Does each spark converge to single-sephirah ownership?

## Acceptance Criteria
- [ ] `BirurimLayer` class with per-sephirah sparse activation
- [ ] `birur_loss` encouraging single-ownership per spark
- [ ] Pillar-aware initialization
- [ ] `SephironV2` flag: `birur=True` (default False)
- [ ] 635+ tests pass
- [ ] Experiment: birur vs no-birur on path_finding + conjunction (3 seeds)
- [ ] Sparse dimension ablation (does 288 win?)
- [ ] Ownership matrix visualization
- [ ] Findings entry

## Why This Is Novel
- **No existing line models per-sephirah feature specialization** — all nodes process full representations
- **288 as architectural constant** has never been tested (unlike 10, 22, 4 which are validated)
- **Sparse activation with learned ownership** is a novel inductive bias from Lurianic Kabbalah
- **Connects F030 → mechanism**: explains WHY Sephiron produces linearly separable representations
- **Distinct from Shevirah-Tikun**: that line models training PHASES; this models the MECHANISM of repair

## Runtime
~30 min CPU (Phase 1), ~75 min total
